client/packages/coldchain/src/Monitoring/ListView/TemperatureChart/useTemperatureChartData.ts

```diff
@@ -0,0 +1,121 @@
+import { useTheme } from '@common/styles';
+import { TemperatureBreachRowFragment, useTemperatureLog } from '../../api';
+import { DateUtils } from '@common/intl';
+import { NumUtils } from '@common/utils';
+
+const MAX_DATA_POINTS = 30;
+
+export type Log = {
+  date: number;
+  sensorId: string;
+  temperature: number | null;
+  temperatureBreach?: TemperatureBreachRowFragment | null;
+};
+
+export type Sensor = {
+  colour: string | undefined;
+  id: string;
+  name: string;
+  location?: string | null;
+  logs: Log[];
+};
+
+export const useTemperatureChartData = () => {
+  const theme = useTheme();
+  const { data, isLoading } = useTemperatureLog.document.list();
+  const sensors: Sensor[] = [];
+  const logs: Log[] = [];
+
+  data?.nodes?.forEach(
+    ({ datetime, temperature, sensor, temperatureBreach }) => {
+      if (!sensor) return;
+      let sensorIndex = sensors.findIndex(s => s.id === sensor.id);
+      if (sensorIndex === -1) {
+        sensors.push({
+          colour:
+            theme.palette.chart.lines[
+              sensors.length % theme.palette.chart.lines.length
+            ],
+          id: sensor.id,
+          name: sensor.name,
+          logs: [],
+          location: sensor.location?.name,
+        });
+        sensorIndex = sensors.length - 1;
+      }
+
+      const date = DateUtils.getDateOrNull(datetime);
+      if (date === null) return;
+
+      logs.push({
+        date: date.getTime(),
+        sensorId: sensor.id,
+        temperature,
+        temperatureBreach,
+      });
+    }
+  );
+  const numOfDataPoints = Math.min(MAX_DATA_POINTS, logs.length);
+  // the fromDate & toDate will come from filters
+  // until that is implemented, we will use the first and last log dates
+  const sortedLogs = logs.sort((a, b) => a.date - b.date);
+  const fromDate = sortedLogs[0]?.date ?? new Date().getTime();
+  const toDate =
+    sortedLogs[sortedLogs.length - 1]?.date ?? new Date().getTime();
+
+  const chartDuration = toDate - fromDate;
+  const periodDuration = chartDuration / numOfDataPoints;
+
+  sensors.forEach(sensor => {
+    sensor.logs = Array.from({
+      length: numOfDataPoints,
+    }).map((_, i) => {
```

Comment:
I really think that this should be done on backend. Similar to the chart for item, front end can specify number of data points and to/from, and the return would be something like:
```
nodes: [
 sensor: SensorNode
 chartData: {
   timestamp
   averageTemperature
   maxTemperature
   minTemperature
 }
 breaches: BreachNode[]
]
```

I think the aggregation should mostly be done in database itself, but I think it's not trivial with diesel-rs, we can probably just copy paste a raw sql query and bind it's output ([raw query something like this](https://github.com/openmsupply/msupply-cold-chain/blob/30e33a6a950494da7a27bbb860cd258c72b23673/src/features/Chart/ChartManager.ts)). Since group by would require diesel update, we probalby ok with raw query for now, in the future I think we can use group by diesel with aggregates and calculate the content of 'with recursive' manually, since it's not [supported by diesel-rs yet](https://github.com/diesel-rs/diesel/issues/356).

Anyways I don't think we need to do this in this PR, but i think it should be in for MVP (for performance)

Reply:
yeah, agree. This issue had taken more time than I wanted already, and the PR size was getting large so I didn't want to add in the backend aggregation in this issue. Better to get a working example, and that helps us define what we want the backend change to do.

Created #2404